shared-bindings/_canio/CAN.c

```diff
@@ -0,0 +1,385 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 Jeff Epler for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/enum.h"
+#include "common-hal/_canio/CAN.h"
+#include "common-hal/_canio/Listener.h"
+#include "shared-bindings/_canio/__init__.h"
+#include "shared-bindings/_canio/CAN.h"
+#include "shared-bindings/_canio/Listener.h"
+#include "shared-bindings/_canio/Match.h"
+#include "shared-bindings/_canio/Message.h"
+#include "shared-bindings/microcontroller/Pin.h"
+
+#include "py/objproperty.h"
+#include "py/runtime.h"
+
+//|
+//| class CAN:
+//|     """CAN bus protocol"""
+//|
+//|     def __init__(self,
+//|             rx: microcontroller.Pin,
```

Comment:
Is there a case for being TX-only?

Reply:
Both STM32 and SAMD51 support 4 modes: normal, silent, external loopback, and silent loopback.  For feature parity with the MicroPython CAN implementation, we need to support all 4 modes.

 * normal mode: this is self explanatory.  Send and receive messages
 * silent (also known as monitor) mode: Sniff a CAN bus while being confident you won't disrupt its operation
 * (external) loopback mode (also could be called "tx only"): super useful when developing CAN transmitter code because it means you can logic probe what your device is transmitting, but you don't need a working CAN transceiver or another device on the CAN bus to send ACK messages.
 * silent (also called internal) loopback mode is useful for performing a "Hot Selftest".

Loopback modes also cause the device to receive its own messages, which is not normally the case.

Silent loopback is the weirdest to wrap my head around, so I'll just paste from the SAM E54 datasheet:
> Internal Loop Back Mode … can be used for a “Hot Selftest”, meaning the CAN can be tested without affecting a running CAN system connected to the pins CAN_TX and CAN_RX. In this mode pin CAN_RX is disconnected from the CAN and pin CAN_TX is held recessive.

(This isn't quite what my code does, because to enter silent mode `tx` is not specified, so the special function isn't activated and the related pin is actually usable as GPIO.)

Additionally, even for silent loopback mode, we need to specify at least one HW pin in order to select the right CAN instance (CAN0 vs CAN1).

I'm open to other ways to specify it, but the way I saw was by making one pin (arbitrarily, tx) optional and making the other aspect controlled by a boolean argument.  If you're still uneasy, mind sketching what you have in mind after reading this explanation?

Reply:
If I choose normal/silent/external loopback based on rx/tx being specified or not, then I still need a bool parameter to specify internal loopback / hot selftest mode.  Is that better than the way it is now?  Each way I can think of writing it, the asymmetry bugs me.

Reply:
You could have the fourth "hot selftest" be when neither TX or RX are provided. However, that is likely to cause more confusion than solve it. I think we either 1) don't support "hot selftest" or 2) add an `enable` property that turns external connection off and on.

Reply:
OK, now one of RX or TX can be omitted, and silent/loopback will be specified separately.  They aren't checked for appropriateness, so for instance you can specify `tx=CAN_TX, rx=None, silent=True` and you will never be able to recieve or transmit anything.

Reply:
changed _AGAIN_, the HW block really wants both TX and RX so both are required.  Also changed the order to match UART